File: p1/arbol.cpp
/**
 * Universidad de La Laguna
 * Escuela Superior de Ingenieria y Tecnologia
 * Inteligencia Artificial
 *
 * @author Juan Rodríguez Suárez. alu0101477596@ull.edu.es
 * @date 28. septiembre .2023
 * @brief Práctica 1 de Inteligencia Artificial. Búsquedas en amplitud y profundidad.
 * @file arbol.cpp Contiene la implementación de la clase Arbol.
 */

#include "arbol.hpp"

/**
 * @brief Genera un nodo dinámicamente con un valor, una distancia para llegar a él y un nodo padre.
 * @param kKey Valor asociado al nodo (representativo).
 * @param kDistancia Distancia para llegar al nodo.
 * @param kPadre Puntero al nodo padre.
 * @return Puntero al nodo generado.
 */
Nodo* Arbol::GenerarNodo(const int kNumeroNodo, const double kDistancia, Nodo* kPadre) {
  Nodo* nodo{new Nodo{kNumeroNodo, kDistancia, kPadre}};
  nodos_generados_.emplace_back(nodo);
  return nodo;
}

/**
 * @brief Realiza un recorrido en amplitud hasta encontrar el valor objetivo.
 * @param kNodoInicial Valor del nodo inicial.
 * @param kNodoFinal Valor del nodo final.
 * @return true sí encontró un camino, false si no.
 */
bool Arbol::BFS(const int kNodoInicial, const int kNodoFinal) {
  // Limpiar el árbol anterior.
  destroy_();
  // Inicializar variables.
  head_ = GenerarNodo(kNodoInicial, 0);
  std::vector<int> sucesores;
  std::queue<Nodo*> almacen;
  Nodo* nodo_actual;
  int numero_nodo_actual;
  double distancia_nodo_actual;
  // Colocamos el nodo inicial en la cola.
  almacen.emplace(head_);
  // Mientras haya nodos en la cola y no se haya alcanzado el nodo final que se ejecute el algoritmo de búsqueda en amplitud.
  while (!almacen.empty()) {
    // Sacamos un nodo de la cola para analizarlo.
    nodo_actual = almacen.front();
    almacen.pop();
    numero_nodo_actual = nodo_actual->GetNumero();
    distancia_nodo_actual = nodo_actual->GetDistancia();
    // Analizando el nodo actual.
    nodos_analizados_.emplace_back(nodo_actual);
    if (numero_nodo_actual == kNodoFinal) {
      nodo_final_ = nodo_actual;
      return true;
    }
    // Observamos los sucesores del nodo actual.
    sucesores = grafo_.GetSucesores(numero_nodo_actual);
    for (auto sucesor : sucesores) {
      // Si un sucesor no está en la rama lo añadimos al árbol.
      if (!nodo_actual->NodoEnRama(sucesor)) {
        almacen.emplace(GenerarNodo(sucesor, distancia_nodo_actual + grafo_.GetDistancia(numero_nodo_actual, sucesor), nodo_actual));
      }
    }
  }
  return false;
}

/**
 * @brief Realiza un recorrido en profundidad hasta encontrar el valor objetivo.
 * @param kNodoInicial Valor del nodo inicial.
 * @param kNodoFinal Valor del nodo final.
 * @return true sí encontró un camino, false si no.
 */
bool Arbol::DFS(const int kNodoInicial, const int kNodoFinal) {
  // Limpiar el árbol anterior.
  destroy_();
  // Inicializar variables.
  head_ = GenerarNodo(kNodoInicial, 0);
  std::vector<int> sucesores;
  std::stack<Nodo*> almacen;
  Nodo* nodo_actual;
  int numero_nodo_actual;
  double distancia_nodo_actual;
  // Colocamos el nodo inicial en la pila.
  almacen.emplace(head_);
  // Mientras haya nodos en la pila y no se haya alcanzado el nodo final que se ejecute el algoritmo de búsqueda en profundidad.
  while (!almacen.empty()) {
    // Sacamos un nodo de la pila para analizarlo.
    nodo_actual = almacen.top();
    almacen.pop();
    numero_nodo_actual = nodo_actual->GetNumero();
    distancia_nodo_actual = nodo_actual->GetDistancia();
    // Analizando el nodo actual.
    nodos_analizados_.emplace_back(nodo_actual);
    if (numero_nodo_actual == kNodoFinal) {
      nodo_final_ = nodo_actual;
      return true;
    }
    // Observamos los sucesores del nodo actual.
    sucesores = grafo_.GetSucesores(numero_nodo_actual);
    for (auto sucesor : sucesores) {
      // Si un sucesor no está en la rama lo añadimos al árbol.
      if (!nodo_actual->NodoEnRama(sucesor)) {
        almacen.emplace(GenerarNodo(sucesor, distancia_nodo_actual + grafo_.GetDistancia(numero_nodo_actual, sucesor), nodo_actual));
      }
    }
  }
  return false;
}

/**
 * @brief Calcula el camino desde el nodo final hasta el inicial y lo devuelve en un formato imprimible.
 * @return Cadena con el camino.
 */
std::string Arbol::GetCamino() const {
  std::string camino;
  Nodo* nodo_actual{nodo_final_};
  while (nodo_actual != nullptr) {
    camino += std::to_string(nodo_actual->GetNumero());
    camino += ">";
    nodo_actual = nodo_actual->GetPadre();
  }
  if (camino.length() > 0) {
    camino.pop_back();
  }
  std::reverse(camino.begin(), camino.end());
  camino += "}";
  camino.insert(camino.begin(), 1, '{');
  return camino;
}

/**
 * @brief Calcula el coste del camino obtenido.
 * @return Coste del camino.
 */
double Arbol::GetDistanciaCamino() const {
  if (nodo_final_ == nullptr) return -1;
  return nodo_final_->GetDistancia();
}

/**
 * @brief Devuelve el conjunto de los nodos generados en un formato imprimible.
 * @return Cadena con los nodos generados.
 */
std::string Arbol::GetGenerados() const {
  std::string generados{'{'};
  for (auto nodo : nodos_generados_) {
    generados += std::to_string(nodo->GetNumero()) + ", ";
  }
  if (generados.length() > 1) {
    for (int i{0}; i < 2; ++i) generados.pop_back();
  }
  generados += "}";
  return generados;
}

/**
 * @brief Devuelve el conjunto de los nodos analizados en un formato imprimible.
 * @return Cadena con los nodos analizados.
 */
std::string Arbol::GetAnalizados() const {
  std::string analizados{'{'};
  for (auto nodo : nodos_analizados_) {
    analizados += std::to_string(nodo->GetNumero()) + ", ";
  }
  if (analizados.length() > 1) {
    for (int i{0}; i < 2; ++i) analizados.pop_back();
  }
  analizados += "}";
  return analizados;
}

/**
 * @brief Limpia toda la memoria empleada en la creación de los nodos y reasigna las variables del Arbol a su valor inicial.
 */
void Arbol::destroy_() {
  head_ = nullptr;
  nodo_final_ = nullptr;
  nodos_analizados_.clear();
  for (auto nodo : nodos_generados_) delete nodo;
  nodos_generados_.clear();
}

/**
 * @brief Destructor de Arbol.
 */
Arbol::~Arbol() {
  destroy_();
}

File: p2/src/arbol.py
"""
Módulo que contiene la clase Arbol y Nodo.

:file: arbol.py
:author: Juan Rodríguez Suárez
:date: 14/10/2023
"""

class Nodo:
  """
  Clase que representa un nodo del arbol de búsqueda.
  """

  def __init__(self, coordenadas, coste_acumulado, coste_heuristico, padre) -> None:
    """
    Inicializa un nodo del arbol de búsqueda.
    
    :param coordenadas: Coordenadas del nodo.
    :param coste_acumulado: Coste acumulado del nodo.
    :param coste_heuristico: Coste heurístico del nodo.
    :param padre: Nodo padre del nodo.
    :return:
    """

    self.coordenadas = coordenadas
    self.coste_acumulado = coste_acumulado
    self.coste_heuristico = coste_heuristico
    self.padre = padre

  def __str__(self) -> str:
    """
    Representación en cadena del nodo.

    :return: Representación en cadena del nodo.
    """

    return self.coordenadas

  def CosteTotal(self) -> int:
    """
    Coste total del nodo, es decir, la suma del coste acumulado hasta ahora y el coste heurístico.

    :return: Coste total del nodo.
    """

    return self.coste_acumulado + self.coste_heuristico

class Arbol:
  """
  Clase que representa el arbol de búsqueda.
  """

  def __init__(self, coordenadas_head, coste_heuristico_head) -> None:
    """
    Inicializa el arbol de búsqueda.

    :param coordenadas_head: Coordenadas del nodo inicial.
    :param coste_heuristico_head: Coste heurístico del nodo inicial.
    :return:
    """

    self.generados = []
    self.analizados = []
    self.final = None
    self.head = self.CrearNodo(coordenadas_head, 0, coste_heuristico_head, None)
  
  def CrearNodo(self, coordenadas, coste_acumulado, coste_heuristico, padre) -> Nodo:
    """
    Crea un nuevo nodo.

    :param coordenadas: Coordenadas del nodo.
    :param coste_acumulado: Coste acumulado del nodo.
    :param coste_heuristico: Coste heurístico del nodo.
    :param padre: Nodo padre del nodo.
    :return: Nuevo nodo.
    """

    nuevo_nodo = Nodo(coordenadas, coste_acumulado, coste_heuristico, padre)
    self.generados.append(nuevo_nodo.coordenadas)
    return nuevo_nodo
  
  def GetCamino(self) -> list:
    """
    Devuelve el camino desde el nodo inicial hasta el nodo final.

    :return: Camino desde el nodo inicial hasta el nodo final.
    """

    camino = []
    nodo_actual = self.final
    while nodo_actual != None:
      camino.append(nodo_actual.coordenadas)
      nodo_actual = nodo_actual.padre
    camino.reverse()
    return camino
    